roshdology123/ToDo-List-MD-labs
# Todo list app

# Pull Request: Todo List App Implementation
## Description

This PR introduces a Flutter-based Todo List application with a clean architecture, robust state management, and essential functionalities for managing tasks effectively.
Key Features

    Reactive State Management: Implemented using GetX for seamless and efficient UI updates.
    Persistent Task Storage: Tasks persist between app sessions using SharedPreferences.
    Core CRUD Operations:
        Create new tasks.
        Edit existing tasks.
        Mark tasks as completed.
        Remove tasks.
    Responsive Design: Ensures a clean and adaptable user experience across devices.
    Unit Testing: Comprehensive tests for key functionalities to ensure robustness.

### Changes Implemented
Features

Create tasks with title, description, due date, and priority.
Edit existing tasks and update fields dynamically.
Mark tasks as completed/uncompleted with a single tap.
Delete tasks via swipe actions.

    Persist tasks locally using SharedPreferences.

### Technical Details

    Models: TodoModel defines the data structure with attributes like title, description, priority, and due date.
    Controllers: TodoController handles the application state and business logic.
    Views:
        TodoView: Displays the list of tasks with dynamic updates.
        AddTaskDetailView: Allows task creation and editing with enhanced validation and UX.
    Unit Testing:
        Verified TodoController functionality for adding, editing, and removing tasks.
        Mocked SharedPreferences for persistence testing.

Video

https://github.com/user-attachments/assets/9322eaea-ea25-4ef0-8c18-ad6ffa4f8983


Testing Checklist
![image](https://github.com/user-attachments/assets/ac8258a6-ea56-49b5-a807-94b893681710)
![image](https://github.com/user-attachments/assets/e835d9a6-c1c1-4476-acdf-a620a0d2eee4)
![image](https://github.com/user-attachments/assets/f66249d4-9f51-4d78-a2fd-ec75ecd3604b)


Manual Testing:

Create tasks with all fields.
Edit tasks and verify updates.
Mark tasks as complete/incomplete.
Swipe to delete tasks.
Validate task persistence across sessions.

        Test app responsiveness on different devices.

How to Test

    Clone the repository.
    Run flutter pub get to install dependencies.
    Execute tests:

flutter test

Run the app:

`flutter run`

Dependency Updates
  get: ^4.6.6
  shared_preferences: ^2.3.3
  animations: ^2.0.11
  intl: ^0.19.0

Potential Future Improvements

Implement task filtering (completed/pending).
Add advanced persistent storage options (e.g., Hive, SQLite).
Integrate recurring tasks functionality.
Add local notifications for due dates.

    Introduce task tagging and categorization.

Reviewer Checklist

- [ ]  Code adheres to Flutter/Dart best practices.
- [ ]  No lint warnings/errors.
- [ ]  UI is clean, consistent, and responsive.
- [ ]  Proper error handling and validations are implemented.

Additional Notes

    Design Philosophy: Simple, clean, and user-focused.
    Architecture: Scalable and modular for future feature expansion.
    Dependencies: Minimal for better maintainability and performance.
